### PR TITLE
CHE-3129 - Project can't be created without Z option

### LINKED ISSUE #3129: Project can't be created without Z option
**Reproduction Steps:**

- set the option
che.docker.volumes_projects_options=
- try to create a project

**Expected behavior:**

The project is created.

**Observed behavior:**

You will encounter a 500 internal error

**Che version:**     5.0.0-M8-SNAPSHOT, rev 13000b066f56d27792e9b625cb0c8742d04e6444   
**OS and version:** Fedora 24 
**Docker version:** 1.12.3  
**Che install:**    Docker container   

**Additional information:**

* Problem started happening recently, didn't happen in an older version of Che: [No]
* Problem can be reliably reproduced, doesn't happen randomly: [Yes]

### PR DESCRIPTION
**Pull request by snjeza**
### What does this PR do?

It fixes #3129

### What issues does this PR fix or reference?

#3129

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

### CODE REVIEW

**Comment by benoitf** [comment]
ci-build

**Comment by tolusha** [comment]
@garagatyi 
Could you have a look pls. A think it is matter of `che.docker.volumes_projects_options` configuration not code changing.

**Comment by l0rd** [comment]
@tolusha I think that #3129 needs code changing (in the default config [the property is set](https://github.com/eclipse/che/blob/b7682ddde70f40597cf208f8296e529fbbe4c6c3/assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/codenvy/che.properties#L137)).

@snjeza I think you need do the same check for property `che.docker.volumes_agent_options` and if you could flag the fields/parameters with `@Nullable` it would be nice.

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/LocalCheInfrastructureProvisioner.java [hunk 1]
```diff
@@ -69,9 +71,15 @@ public void provision(Environment envConfig, CheServicesEnvironmentImpl internal
         // add bind-mount volume for projects in a workspace
         String projectFolderVolume;
         try {
-            projectFolderVolume = format("%s:%s:%s",
+            if (Strings.isNullOrEmpty(volumesOptions)) {
```
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/LocalCheInfrastructureProvisioner.java]: Let's form suffix in constructor. Then you won't need this clause in main code.

**Review by garagatyi — CHANGES_REQUESTED** [review]
@snjeza Nice catch!
@tolusha No, code have to be changed in any case.
@l0rd Agree with you, Mario!

@snjeza Please consider comments of Mario and mine. Please also rename volumeOptions argument of constructor to projectsVolumeOptions

**Comment by snjeza** [comment]
I have updated the patch.

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/LocalCheInfrastructureProvisioner.java [hunk 2]
```diff
@@ -35,19 +38,23 @@
     private final WorkspaceFolderPathProvider workspaceFolderPathProvider;
     private final WindowsPathEscaper          pathEscaper;
     private final String                      projectFolderPath;
-	private final String                      volumesOptions;
+	private final String                      projectsVolumeOptions;
```
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/LocalCheInfrastructureProvisioner.java]: Please replace tab character with spaces. Someone put it here but it is not allowed for indentation in Che.

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/DockerExtConfBindingProvider.java [hunk 3]
```diff
@@ -76,7 +79,11 @@ public String get() {
     }
 
     private String getTargetOptions(final String path) {
-        return path + CONTAINER_TARGET + ":" + volumeOptions;
+        if (!Strings.isNullOrEmpty(agentVolumeOptions)) {
```
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/DockerExtConfBindingProvider.java]: Please create constructor and put this code there as you did in LocalCheInfrastructureProvisioner

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/TerminalVolumeProvider.java [hunk 4]
```diff
@@ -64,7 +67,11 @@ public String get() {
     }
 
     private String getTargetOptions(final String path) {
-        return path + CONTAINER_TARGET + ":" + volumeOptions;
+        if (!Strings.isNullOrEmpty(agentVolumeOptions)) {
```
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/TerminalVolumeProvider.java]: Please create constructor and put this code there as you did in LocalCheInfrastructureProvisioner

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/WsAgentVolumeProvider.java [hunk 5]
```diff
@@ -68,6 +71,10 @@ public String get() {
     }
 
     private String getTargetOptions(final String path) {
-        return path + CONTAINER_TARGET + ":" + volumeOptions;
+        if (!Strings.isNullOrEmpty(agentVolumeOptions)) {
```
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/WsAgentVolumeProvider.java]: Please create constructor and put this code there as you did in LocalCheInfrastructureProvisioner

**Review by garagatyi — CHANGES_REQUESTED** [review]

**Comment by snjeza** [comment]
The patch updated.

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/DockerExtConfBindingProvider.java [hunk 6]
```diff
@@ -43,8 +46,16 @@
     private static final Logger LOG              = LoggerFactory.getLogger(DockerExtConfBindingProvider.class);
 
     @Inject
-    @Named("che.docker.volumes_agent_options")
-    private String volumeOptions;
+    @Nullable @Named("che.docker.volumes_agent_options")
+    private String agentVolumeOptions;
+
+    public DockerExtConfBindingProvider() {
+        if (!Strings.isNullOrEmpty(agentVolumeOptions)) {
```
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/DockerExtConfBindingProvider.java]: Constructor is called before field injection in Guice. So here we will have null value always. Can you replace field injection with a constructor injection.

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/TerminalVolumeProvider.java [hunk 7]
```diff
@@ -43,8 +46,16 @@
     private String terminalArchivePath;
 
     @Inject
-    @Named("che.docker.volumes_agent_options")
-    private String volumeOptions;
+    @Nullable @Named("che.docker.volumes_agent_options")
+    private String agentVolumeOptions;
+
+    public TerminalVolumeProvider() {
+        if (!Strings.isNullOrEmpty(agentVolumeOptions)) {
```
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/TerminalVolumeProvider.java:51]: Constructor is called before field injection in Guice. So here we will have null value always. Can you replace field injection with a constructor injection.

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/WsAgentVolumeProvider.java [hunk 8]
```diff
@@ -47,8 +50,16 @@
     private String wsAgentArchivePath;
 
     @Inject
-    @Named("che.docker.volumes_agent_options")
-    private String volumeOptions;
+    @Nullable @Named("che.docker.volumes_agent_options")
+    private String agentVolumeOptions;
+
+    public WsAgentVolumeProvider() {
+        if (!Strings.isNullOrEmpty(agentVolumeOptions)) {
```
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/WsAgentVolumeProvider.java:55]: Constructor is called before field injection in Guice. So here we will have null value always. Can you replace field injection with a constructor injection.

**Review by garagatyi — CHANGES_REQUESTED** [review]
Sorry that still asking you to change your PR, but there are few minor changes to be made.

**Comment by snjeza** [comment]
The patch updated.
I wasn't able to inject the che.docker.volumes_agent_options option to the DockerExtConfBindingProvider constructor because DockerExtServerModule created it with the empty constructor.
Che doesn't start when I try to bind DockerExtConfBindingProvider to DockerExtServerModule.

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/DockerExtConfBindingProvider.java [hunk 9]
```diff
@@ -43,8 +46,8 @@
     private static final Logger LOG              = LoggerFactory.getLogger(DockerExtConfBindingProvider.class);
 
     @Inject
-    @Named("che.docker.volumes_agent_options")
-    private String volumeOptions;
+    @Nullable @Named("che.docker.volumes_agent_options")
```
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ext/provider/DockerExtConfBindingProvider.java:49]: OK, then leave it as it is now, I'll fix that in one of my next PRs.

**Review by garagatyi — APPROVED** [review]
Good job, Snjezana!

**Comment by sleshchenko** [comment]
Looks like DockerExtConfBindingProvider shouldn't work correctly because instance is created with default constructor and requestInjection method is not invoked for it. Please make sure that it works correctly

**Comment by snjeza** [comment]
> Looks like DockerExtConfBindingProvider shouldn't work correctly because instance is created 
> with default constructor and requestInjection method is not invoked for it. Please make sure 
> that it works correctly

That's why I haven't injected the che.docker.volumes_agent_options option to the DockerExtConfBindingProvider constructor.
I have tested the build with and without the Z option. It works properly.

**Comment by garagatyi** [comment]
If instance of class is created manually and injection is not requested it should not work if I understand how it works. Can you describe how did you test that?

**Comment by snjeza** [comment]
I have created che-server, started it with different che.properies, created a project, reviewed mounted volumes using "docker inspect". Everything works as expected.

**Comment by garagatyi** [comment]
@sleshchenko correct me if I'm mistaken but looks like ws-agent configuration works in a different way.

**Comment by snjeza** [comment]
When excluding the Z option, the volumes are mounted in the following way:

```
 "Mounts": [
            {
                "Source": "/data/workspaces/wksp-8t9p",
                "Destination": "/projects",
                "Mode": "",
                "RW": true,
                "Propagation": "rprivate"
            },
            {
                "Source": "/tmp/data/lib/linux_amd64/terminal",
                "Destination": "/mnt/che/terminal",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            },
            {
                "Source": "/tmp/data/lib/ws-agent.tar.gz",
                "Destination": "/mnt/che/ws-agent.tar.gz",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            }
        ],
```

**Comment by sleshchenko** [comment]
@snjeza Please create folder plugin-conf in CHE_LOCAL_CONF_DIR. As far as I understand it can be empty. And then mounts should include extra one with `"Destination": /mnt/che/conf""`. Please make ensure that it is mounted with 'Z' suffix when configuration contains it. If no, please try to add `requestInjection(extConfBindingProvider);` in DockerExtServerModule

**Comment by snjeza** [comment]
The following are the mounted volumes:

```
 "Mounts": [
            {
                "Source": "/data/workspaces/wksp-8t9p",
                "Destination": "/projects",
                "Mode": "",
                "RW": true,
                "Propagation": "rprivate"
            },
            {
                "Source": "/tmp/data/lib/linux_amd64/terminal",
                "Destination": "/mnt/che/terminal",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            },
            {
                "Source": "/tmp/data/lib/ws-agent.tar.gz",
                "Destination": "/mnt/che/ws-agent.tar.gz",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            },
            {
                "Source": "/conf/plugin-conf",
                "Destination": "/mnt/che/conf",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            }
        ],
```

with the Z option

```
"Mounts": [
            {
                "Source": "/data/workspaces/wksp-jvuc",
                "Destination": "/projects",
                "Mode": "Z",
                "RW": true,
                "Propagation": "rprivate"
            },
            {
                "Source": "/tmp/lib/linux_amd64/terminal",
                "Destination": "/mnt/che/terminal",
                "Mode": "ro,Z",
                "RW": false,
                "Propagation": "rprivate"
            },
            {
                "Source": "/data/conf/plugin-conf",
                "Destination": "/mnt/che/conf",
                "Mode": "ro,Z",
                "RW": false,
                "Propagation": "rprivate"
            },
            {
                "Source": "/tmp/lib/ws-agent.tar.gz",
                "Destination": "/mnt/che/ws-agent.tar.gz",
                "Mode": "ro,Z",
                "RW": false,
                "Propagation": "rprivate"
            }
        ],
```

**Comment by garagatyi** [comment]
It is interesting. We debugged DockerExtConfBindingProvider usage and guice injected dependencies between 1st and 2nd calls of get method of this class. Sorry that we asked you to check more and more but we missed that Guice behavior with singleton classes.

**Review by sleshchenko — APPROVED** [review]

**Comment by garagatyi** [comment]
ci-build

**Review by vparfonov — APPROVED** [review]

**Pull request merged**